### PR TITLE
Obdscan

### PR DESCRIPTION
**Pull request by AlexAshs**
A tool, that analyzes packages based on the OnBoardDiagnostics-Protocol on the CAN-Bus.
It offers the output of current DiagnosticTroubleCodes,
scanning for supported Parameter IDentifiers and
unsupported(hidden) PIDs

### CODE REVIEW

File: binascii [hunk 1]
```diff
@@ -0,0 +1,285 @@
+%!PS-Adobe-3.0
```
Comment (polybassa) [inline binascii]: Please remove this file and the file `python_can`, `socket`, `threading` from your PR. 
Please undo the modification to `isotpscanner.py`.

File: scapy/contrib/automotive/obd/scanner.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+# Copyright (C) Andreas Korb <andreas.korb@e-mundo.de>
+# Copyright (C) Friedrich Feigel <friedrich.feigel@e-mundo.de>
+# This program is published under a GPLv2 license
+
+import sys
+from scapy.compat import chb
+from scapy.contrib.automotive.obd.obd import *
+
+
+def _supported_id_numbers(socket, timeout, service_class, id_name, verbose):
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: could you please add some documentation for the parameters

File: scapy/contrib/automotive/obd/scanner.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    if resp is not None:
+        return bytes(resp)
+
+
+def obd_scan(socket, timeout=0.1, supported_ids=False, unsupported_ids=False, verbose=False):
+    dtc = dict()
+    supported = dict()
+    unsupported = dict()
+
+    print("\nStarting OBD-Scan...")
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: could you make the print statements dependant on `verbose`

File: scapy/contrib/automotive/obd/scanner.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+
+def _scan_dtc_service(socket, timeout, service_class, verbose):
+    req = OBD()/service_class()
+    resp = socket.sr1(req, timeout=timeout, verbose=verbose)
+    if resp is not None:
+        return bytes(resp)
+
+
+def obd_scan(socket, timeout=0.1, supported_ids=False, unsupported_ids=False, verbose=False):
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: please document this function and its arguments

File: scapy/contrib/automotive/obd/scanner.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    # On-board monitoring testresults for non-continuously monitored systems
+    supported_ids_s06 = _supported_id_numbers(socket, timeout, OBD_S06, 'mid', verbose)
+    # Control of on-board system, test or component
+    supported_ids_s08 = _supported_id_numbers(socket, timeout, OBD_S08, 'tid', verbose)
+    # On-board monitoring test results for non-continuously monitored systems
+    supported_ids_s09 = _supported_id_numbers(socket, timeout, OBD_S09, 'iid', verbose)
+
+    if supported_ids:
+        print("\nScanning supported Parameter IDs")
+        supported[1] = _scan_id_service(socket, timeout, OBD_S01, supported_ids_s01, verbose)
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: What about `OBD_S02`?

File: scapy/contrib/automotive/obd/scanner.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        supported_ids_req = None
+
+        all_supported_in_range = getattr(resp.data_records[0], supported_prop)
+
+        for supported in all_supported_in_range:
+            id_number = int(supported[-2:], 16)
+            supported_id_numbers.add(id_number)
+
+            # check whether supported id is for next id range
+            if id_number % 20 == 0:
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: this should be 0x20, shouldn't it?

File: scapy/tools/automotive/obdscanner.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    sys.exit(0)
+
+
+def usage():
+    print('''usage:\tobdscanner [-i|--interface] [-c|--channel] [-s|--source] [-d|--destination] [-b|--bitrate] 
+                                [-h|--help] [-t|--timeout] [-r|--supported] [-u|--unsupported] [-v|--verbose]\n
+    Scan for all possible obd service classes and their subfunctions.\n
+    optional arguments:
+    -c, --channel               python-can channel or Linux SocketCAN interface name
+    -s, --source                ISOTP-socket source id (hex)
```
Comment (polybassa) [inline scapy/tools/automotive/obdscanner.py]: could you please add default values for source and destination

File: scapy/tools/automotive/obdscanner.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    try:
+        csock = CANSocket(iface=scan_interface)
+    except Exception as e:
+        usage()
+        print("\nSocket couldn't be created. Check your arguments.\n",
+              file=sys.stderr)
+        print(e, file=sys.stderr)
+        sys.exit(-1)
+
+    isock = ISOTPSocket(csock, source, destination, basecls=OBD, padding=True)
```
Comment (polybassa) [inline scapy/tools/automotive/obdscanner.py]: could you please use 
```
with  ISOTPSocket(csock, source, destination, basecls=OBD, padding=True) as isock:
    ...
    ...
```

File: scapy/tools/automotive/obdscanner.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+    signal.signal(signal.SIGINT, signal_handler)
+
+    obd_scan(isock, timeout, supported, unsupported, verbose)
+
+    csock.close()
+    isock.close()
```
Comment (polybassa) [inline scapy/tools/automotive/obdscanner.py]: not necessary if you use with

File: test/tools/obdscanner.uts [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+expected_output = [bytes("Service 3:\nb'C\\x01\\x00\\x00'", 'utf-8'),
+                   bytes("Service 1:\n{1: b'A\\x01\\x00\\x00\\x00\\x00'}", 'utf-8')]
+print(std_out)
+for out in expected_output:
+    assert plain_str(out) in plain_str(std_out)
+
+
+= Cleanup
+responder.stop()
+if 0 != call("sudo ip link delete vcan0", shell=True):
```
Comment (polybassa) [inline test/tools/obdscanner.uts]: nice way to kill the sockets

**Comment by gpotter2** [comment]
Please rebase against master. We have implemented some new stricter checks concerning the docstrings :-)

File: binascii (hunk 1 above)
Comment (AlexAshs) [reply binascii]: Deleted.
Do you mean actual changes to `ìsotpscanner.py`?
I am probably missing something, but if I compare it to your last version of the file, I don't see any alterations.

Comment (polybassa) [reply binascii]: You changed the file permissions, I think

Comment (AlexAshs) [reply binascii]: Yes, you are right!
I changed it back to 644

File: scapy/contrib/automotive/obd/scanner.py (hunk 2 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: Done.

File: scapy/contrib/automotive/obd/scanner.py (hunk 4 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: Done.

File: scapy/contrib/automotive/obd/scanner.py (hunk 5 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: On consultation predecessor @akorb stated, that this service, which is being used to monitor freeze frames, is hardly being used, which is why, for us it didn't seem to be worth the effort.

What is your opinion about that? 
Should it be implemented as well?

File: scapy/contrib/automotive/obd/scanner.py (hunk 6 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: Absolutely.
It contains the Bits indicating, whether the next 32 PIDs are supported or not.

File: scapy/contrib/automotive/obd/scanner.py (hunk 3 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: Done.

File: scapy/tools/automotive/obdscanner.py (hunk 7 above)
Comment (AlexAshs) [reply scapy/tools/automotive/obdscanner.py]: I've set the destination to the OBD-Broadcast-Domain.
Should they be non-mandatory, then?

Comment (polybassa) [reply scapy/tools/automotive/obdscanner.py]: Yes, I would like to have source and destination as optional arguments. This makes the default usage faster

File: scapy/contrib/automotive/obd/scanner.py (hunk 5 above)
Comment (polybassa) [reply scapy/contrib/automotive/obd/scanner.py]: no, it's fine

File: scapy/tools/automotive/obdscanner.py (hunk 7 above)
Comment (AlexAshs) [reply scapy/tools/automotive/obdscanner.py]: Ok.

File: scapy/tools/automotive/obdscanner.py (hunk 8 above)
Comment (AlexAshs) [reply scapy/tools/automotive/obdscanner.py]: Done.

File: scapy/tools/automotive/obdscanner.py (hunk 9 above)
Comment (AlexAshs) [reply scapy/tools/automotive/obdscanner.py]: Removed.

File: test/tools/obdscanner.uts (hunk 10 above)
Comment (AlexAshs) [reply test/tools/obdscanner.uts]: Thank you :)

File: scapy/contrib/automotive/obd/scanner.py (hunk 2 above)
Comment (polybassa) [reply scapy/contrib/automotive/obd/scanner.py]: Could you please use the same documentation style like other files. Here is an example: https://github.com/secdev/scapy/blob/a864193915adff0e00bb5a9bcd3f2ffd62f9e43d/scapy/contrib/isotp.py#L1830-L1854

File: scapy/contrib/automotive/obd/scanner.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+            supported_id_numbers.add(id_number)
+
+            # check whether supported id is for next id range
+            if id_number % 0x20 == 0:
+                supported_ids_req = OBD()/service_class(chb(id_number))
+
+    return supported_id_numbers
+
+
+def _scan_id_service(socket, timeout, service_class, id_numbers, verbose):
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: Please document this function

File: scapy/contrib/automotive/obd/scanner.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        # assemble request packet
+        pkt = OBD()/service_class(id_byte)
+        resp = socket.sr1(pkt, timeout=timeout, verbose=verbose)
+
+        if resp is not None:
+            data[id_number] = bytes(resp)
+    return data
+
+
+def _scan_dtc_service(socket, timeout, service_class, verbose):
```
Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: Please document this function

Comment (polybassa) [inline scapy/contrib/automotive/obd/scanner.py]: On thing to consider for your scan and supported functions. The arguments `socket`, `timeout` and `verbose` are just passed through to be used on the call `resp = socket.sr1(pkt, timeout=timeout, verbose=verbose)`.

What about this:
```
def _scan_dtc_service(_sr1, service_class):
    req = OBD()/service_class()
    resp = _sr1(req)
    if resp is not None:
        return bytes(resp)

...
dtc[3] = _scan_dtc_service(lambda pkt: socket.sr1(pkt, timeout=timeout, verbose=verbose), OBD_S03)

# OR
sr1_func = lambda pkt: socket.sr1(pkt, timeout=timeout, verbose=verbose)
dtc[3] = _scan_dtc_service(sr1_func, OBD_S03)
...
supported_ids_s08 = _supported_id_numbers(sr1_func, OBD_S08, 'tid')

```

Just consider it to remove function arguments... If you don't like this.. ignore it.

File: scapy/contrib/automotive/obd/scanner.py (hunk 2 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: Oh sorry, of course!

File: scapy/contrib/automotive/obd/scanner.py (hunk 12 above)
Comment (AlexAshs) [reply scapy/contrib/automotive/obd/scanner.py]: Yes, good point.
In my opinion my current approach seems kind of redundant, while your proposal, though being more elegant, seems to me a bit uneasy to the eye on first sight.

**Comment by AlexAshs** [comment]
Hi @gpotter2 , I'm all new to this, so maybe this question is foolish:

How can I increase the code coverage, because I have added unit tests for the code I provided, which cover my code to a degree, that should not drag down the whole repository by 0.79%.

However it seems to me, the problem is with some other files, that have less coverage hits, than they used to.

**Comment by gpotter2** [comment]
Hi,

There are sadly a lot of random in the coverage. We get weird decreases a lot in files that were clearly not changed.

We'll ignore that when checking the PR

**Comment by AlexAshs** [comment]
Oh thanks, that would be great :)!

**Comment by AlexAshs** [comment]
@polybassa Hi, I am sorry to bother you again, but because of a rebase your review was automatically dismissed (or maybe it was because of something else, but I did not dismiss it myself), which is why there's still a review needed.
Could you approve the pull request once more?

File: scapy/contrib/automotive/obd/scanner.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
+#! /usr/bin/env python
+# flake8: noqa: F405
```
Comment (gpotter2) [inline scapy/contrib/automotive/obd/scanner.py]: I understand that all OBD_S... are a pain to import without a star import.

**Pull request merged**